posthog/temporal/workflows/logger.py

```diff
@@ -0,0 +1,272 @@
+    except RuntimeError:
+        return None
+    else:
+        workflow_id = activity_info.workflow_id
+        workflow_type = activity_info.workflow_type
+        workflow_run_id = activity_info.workflow_run_id
+        attempt = activity_info.attempt
+
+    return (workflow_id, workflow_type, workflow_run_id, attempt)
+
+
+def attempt_to_fetch_workflow_info() -> Info | None:
+    """Fetch Workflow information from Temporal.
+
+    Returns:
+        None if calling outside a Workflow, else the relevant Info.
+    """
+    try:
+        workflow_info = temporalio.workflow.info()
+    except RuntimeError:
+        return None
+    else:
+        workflow_id = workflow_info.workflow_id
+        workflow_type = workflow_info.workflow_type
+        workflow_run_id = workflow_info.run_id
+        attempt = workflow_info.attempt
+
+    return (workflow_id, workflow_type, workflow_run_id, attempt)
+
+
+class KafkaLogProducerFromQueue:
+    """Produce log messages to Kafka by getting them from a queue.
+
+    This KafkaLogProducerFromQueue was designed to ingest logs into the ClickHouse log_entries table.
+    For this reason, the messages we produce to Kafka are serialized as JSON in the schema expected by
+    the log_entries table. Eventually, we could de-couple this producer from the table schema, but
+    schema changes are rare in ClickHouse, and for now we are only using this for logs, so the tight
+    coupling is preferred over the extra complexity of de-coupling this producer.
+
+    Attributes:
+        queue: The queue we are listening to get log event_dicts to serialize and produce.
+        topic: The topic to produce to. This should be left to the default KAFKA_LOG_ENTRIES.
+        key: The key for Kafka partitioning. Default to None for random partition.
+        producer: Optionally, bring your own aiokafka.AIOKafkaProducer. This is mostly here for testing.
+    """
+
+    def __init__(
+        self,
+        queue: asyncio.Queue,
+        topic: str = KAFKA_LOG_ENTRIES,
+        key: str | None = None,
+        producer: aiokafka.AIOKafkaProducer | None = None,
+    ):
+        self.queue = queue
+        self.topic = topic
+        self.key = key
+        self.producer = (
+            producer
+            if producer is not None
+            else aiokafka.AIOKafkaProducer(
+                bootstrap_servers=settings.KAFKA_HOSTS,
+                security_protocol=settings.KAFKA_SECURITY_PROTOCOL or "PLAINTEXT",
+                acks="all",
+                api_version="2.5.0",
+            )
+        )
+
+    async def listen(self):
+        """Listen to messages in queue and produce them to Kafka as they come.
+
+        This is designed to be ran as an asyncio.Task, as it will wait forever for the queue
+        to have messages.
+        """
+        await self.producer.start()
+        try:
+            while True:
+                msg = await self.queue.get()
+                await self.produce(msg)
```

Comment:
So if the Brokers are unhealthy (or something) and this throws, it seems like our Kafka listen task is forever dead and there is recovery or shutdown?

And if so, I guess we'd lose logs and slowly OOM via the `asyncio.Queue(maxsize=-1)`?

I don't think it needs to be super robust or perfect but it does make me a little nervous that there's no real recovery (or I'm missing it!).

Reply:
Still no recovery but at least we will not stop consuming from the queue.

Reply:
Perhaps we should also ignore any errors when initializing the producer, just in case kafka is unhealthy right as we are coming up.

Reply:
Thank you for making our code more reliable!